crates/fuel-core/src/graphql_api/storage/coins.rs

```diff
@@ -8,21 +8,221 @@ use fuel_core_storage::{
     structured_storage::TableWithBlueprint,
     Mappable,
 };
-use fuel_core_types::fuel_tx::{
-    Address,
-    TxId,
-    UtxoId,
+use fuel_core_types::{
+    entities::{
+        coins::coin::Coin,
+        Message,
+    },
+    fuel_tx::{
+        Address,
+        AssetId,
+        TxId,
+        UtxoId,
+    },
+    fuel_types::Nonce,
 };
 
+use crate::graphql_api::indexation;
+
+use self::indexation::{
+    coins_to_spend::{
+        NON_RETRYABLE_BYTE,
+        RETRYABLE_BYTE,
+    },
+    error::IndexationError,
+};
+
+const AMOUNT_SIZE: usize = size_of::<u64>();
+const UTXO_ID_SIZE: usize = size_of::<UtxoId>();
+const RETRYABLE_FLAG_SIZE: usize = size_of::<u8>();
+
 // TODO: Reuse `fuel_vm::storage::double_key` macro.
 pub fn owner_coin_id_key(owner: &Address, coin_id: &UtxoId) -> OwnedCoinKey {
-    let mut default = [0u8; Address::LEN + TxId::LEN + 2];
+    let mut default = [0u8; Address::LEN + UTXO_ID_SIZE];
     default[0..Address::LEN].copy_from_slice(owner.as_ref());
-    let utxo_id_bytes: [u8; TxId::LEN + 2] = utxo_id_to_bytes(coin_id);
+    let utxo_id_bytes: [u8; UTXO_ID_SIZE] = utxo_id_to_bytes(coin_id);
     default[Address::LEN..].copy_from_slice(utxo_id_bytes.as_ref());
     default
 }
 
+/// The storage table for the index of coins to spend.
+
+// In the implementation of getters we use the explicit panic with the message (`expect`)
+// when the key is malformed (incorrect length). This is a bit of a code smell, but it's
+// consistent with how the `double_key!` macro works. We should consider refactoring this
+// in the future.
+pub struct CoinsToSpendIndex;
+
+impl Mappable for CoinsToSpendIndex {
+    type Key = Self::OwnedKey;
+    type OwnedKey = CoinsToSpendIndexKey;
+    type Value = Self::OwnedValue;
+    type OwnedValue = IndexedCoinType;
+}
+
+impl TableWithBlueprint for CoinsToSpendIndex {
+    type Blueprint = Plain<Raw, Raw>;
```

Comment:
nit: We can embed all conversations into the codec like we did with `Manual<OwnedTransactionIndexKey>`. In this case, from the usage perspective, you will always work with nice types with fields instead of vectors or slices.

Reply:
See here: https://github.com/FuelLabs/fuel-core/pull/2463#discussion_r1883811058